### PR TITLE
server/node: get started

### PR DESCRIPTION
**Pull request by srenatus**
This includes:
- update deps, auto-format the sources
- wiring of opa-typescript SDK into the express (server/node) backend
- a basic policy for its API services
- a set of basic Hurl API tests
- a docker-compose file and a github workflow so that
  - the specified server is built, started together with OPA, and the Hurl tests are run

### CODE REVIEW

File: .github/workflows/pull-request.yaml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  cancel-in-progress: true
+
+jobs:
+  test-e2e:
+    name: Test E2E
+    runs-on: ubuntu-22.04
+    strategy:
+      fail-fast: false
+      matrix:
+        server: [node] #TODO: java, csharp
```
Comment (srenatus) [inline .github/workflows/pull-request.yaml:20]: ℹ️ The next service that's ready would be added here, (1.)

File: docker-compose.yaml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    ports:
+      - "4000:4000"
+    profiles:
+      - node
+
+  # TODO
+  server-java:
+    build: server/java/
+    profiles:
+      - java
```
Comment (srenatus) [inline docker-compose.yaml:17]: ...and here (2.)...

File: server/node/Dockerfile
Comment (srenatus) [inline server/node/Dockerfile:1]: ...and would need to bring its own Dockerfile (3.)

The service needs to be able to read the endpoint of OPA to talk to from an environment variable, like `OPA_HOST` in the server/node example

**Review by srenatus — COMMENTED** [review]
Some notes on how to add the next service to the CI setup.

**Review by philipaconrad — APPROVED** [review]

**Comment by philipaconrad** [comment]
@srenatus I am apparently not authorized to merge?

-----
image

**Comment by srenatus** [comment]
@philipaconrad fixed?

File: docker-compose.yaml [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    build: server/node/
+    environment:
+      - OPA_URL=http://opa:8181
+    ports:
+      - "4000:4000"
+    profiles:
+      - node
+
+  # TODO
+  server-java:
```
Comment (charlesdaniels) [inline docker-compose.yaml:14]: I wonder if there is a convenient way with docker-compose to have several different possible server implementations that we swap between in a mutually exclusive way? I guess we could just comment and un-comment.

File: policies/tickets.rego [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+      "alice": ["admin"],
+      "bob": ["reader"],
+      "ceasar": ["reader", "resolver"],
+    },
+    "hooli": {
+      "dylan": ["admin"],
+    }
+}
+
+default allow := false
```
Comment (charlesdaniels) [inline policies/tickets.rego:16]: I have been finding it useful for debugging to have a policy entrypoint that echos back whatever `input` was.

File: server/node/src/authz.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    this.name = "UnauthorizedError";
+  }
+}
+
+export class Authorizer {
+  constructor(endpoint) {
+    this.opa = new OPA(endpoint);
+  }
+
+  async authorized(path, input, { auth: { tenant, subject: user } }) {
```
Comment (charlesdaniels) [inline server/node/src/authz.js:15]: I go back and forth on weather I think the authz should just be a boolean allow/deny or if it should be some kind of exception. On the Java side, the API server does ultimately throw an unauthorized exception that gets consumed by the server framework.

File: tests/api/basic.hurl [hunk 6]
```diff
@@ -0,0 +1,66 @@
+GET http://{{host}}:4000/api/tickets/0
```
Comment (charlesdaniels) [inline tests/api/basic.hurl:1]: This looks slick. I have heard of HURL before. I will have to learn it ...

File: docker-compose.yaml (hunk 3 above)
Comment (srenatus) [reply docker-compose.yaml:14]: That's what the --profile switch allows doing! We also need to set VARIANT.

File: server/node/Dockerfile
Comment (charlesdaniels) [reply server/node/Dockerfile:1]: Some initial searching leads me to believe that there is not a particularly clean way to enable or disable a specific docker-compose service based on environment variables. One option could be to just start them all on different ports and then use `OPA_HOST` like you said to have the frontend choose between them. These are trivial enough that's probably sufficient.

File: docker-compose.yaml (hunk 3 above)
Comment (charlesdaniels) [reply docker-compose.yaml:14]: `--profile stringArray        Specify a profile to enable`

What an illuminating help message :roll_eyes: 

https://docs.docker.com/compose/profiles/

This looks like exactly what we want. TIL that docker compose supports this. Surprised that neither some normal google-fu nor bing chat found this.

**Comment by charlesdaniels** [comment]
FYI I am rebasing my eopa-java-sdk branch on the upstream for this repo off of your new `main`. We may have to do some cleanup on that later.

**Pull request merged**